File: go/pkg/audio/codec/ogg/opus_writer.go
package ogg

import (
	"crypto/rand"
	"encoding/binary"
	"errors"
	"io"
	"sync"
	"time"

	"github.com/haivivi/giztoy/go/pkg/audio/codec/opus"
)

const (
	pageHeaderTypeContinuation = 0x00
	pageHeaderTypeBOS          = 0x02
	pageHeaderTypeEOS          = 0x04
	defaultPreSkip             = 3840 // RFC 7845 §5.1 recommends 80ms pre-skip (3840 samples at 48kHz)
	idPageSignature            = "OpusHead"
	commentPageSignature       = "OpusTags"
	pageHeaderSignature        = "OggS"
	pageHeaderSize             = 27
)

var (
	// ErrWriterClosed is returned when writing to a closed writer.
	ErrWriterClosed = errors.New("ogg: writer is closed")
	// ErrNilWriter is returned when the underlying writer is nil.
	ErrNilWriter = errors.New("ogg: nil writer")
	// ErrInvalidSerialNo is returned when an invalid serial number is used.
	ErrInvalidSerialNo = errors.New("ogg: invalid serial number")
	// ErrStreamEnded is returned when writing to an ended stream.
	ErrStreamEnded = errors.New("ogg: stream already ended")
)

// opusStream represents a single Opus stream within the OGG container.
type opusStream struct {
	serialNo   int32
	sampleRate uint32
	channels   uint16
	granule    int64
	pageIndex  uint32
	ended      bool
}

// OpusWriter writes Opus frames to an OGG container.
// It supports multiple streams with different serial numbers.
type OpusWriter struct {
	mu            sync.Mutex
	w             io.Writer
	streams       map[int32]*opusStream
	defaultStream int32
	checksumTable *[256]uint32
	closed        bool
}

// NewOpusWriter creates a new OGG Opus writer with a default stream.
// The default stream uses a random serial number.
func NewOpusWriter(w io.Writer, sampleRate, channels int) (*OpusWriter, error) {
	if w == nil {
		return nil, ErrNilWriter
	}

	ow := &OpusWriter{
		w:             w,
		streams:       make(map[int32]*opusStream),
		checksumTable: generateChecksumTable(),
	}

	// Create default stream
	serialNo := ow.StreamBegin(sampleRate, channels)
	ow.defaultStream = serialNo

	return ow, nil
}

// StreamBegin creates a new stream with the given sample rate and channels.
// Returns the serial number assigned to this stream.
func (w *OpusWriter) StreamBegin(sampleRate, channels int) int32 {
	w.mu.Lock()
	defer w.mu.Unlock()

	serialNo := w.generateSerialNo()
	stream := &opusStream{
		serialNo:   serialNo,
		sampleRate: uint32(sampleRate),
		channels:   uint16(channels),
		granule:    0,
		pageIndex:  0,
	}
	w.streams[serialNo] = stream

	// Write headers for this stream
	w.writeStreamHeaders(stream)

	return serialNo
}

// generateSerialNo generates a random serial number not already in use.
func (w *OpusWriter) generateSerialNo() int32 {
	for {
		var serialNo int32
		if err := binary.Read(rand.Reader, binary.LittleEndian, &serialNo); err != nil {
			// Fallback to simple increment
			serialNo = int32(len(w.streams) + 1)
		}
		if _, exists := w.streams[serialNo]; !exists {
			return serialNo
		}
	}
}

// writeStreamHeaders writes the OpusHead and OpusTags headers for a stream.
func (w *OpusWriter) writeStreamHeaders(stream *opusStream) error {
	// ID Header (OpusHead)
	idHeader := make([]byte, 19)
	copy(idHeader[0:], idPageSignature)                          // "OpusHead"
	idHeader[8] = 1                                               // Version
	idHeader[9] = uint8(stream.channels)                          // Channel count
	binary.LittleEndian.PutUint16(idHeader[10:], defaultPreSkip)  // Pre-skip
	binary.LittleEndian.PutUint32(idHeader[12:], stream.sampleRate) // Sample rate
	binary.LittleEndian.PutUint16(idHeader[16:], 0)               // Output gain
	idHeader[18] = 0                                               // Channel mapping

	page := w.createPage(idHeader, pageHeaderTypeBOS, 0, stream.pageIndex, stream.serialNo)
	if _, err := w.w.Write(page); err != nil {
		return err
	}
	stream.pageIndex++

	// Comment Header (OpusTags)
	commentHeader := make([]byte, 22)
	copy(commentHeader[0:], commentPageSignature)         // "OpusTags"
	binary.LittleEndian.PutUint32(commentHeader[8:], 6)   // Vendor length
	copy(commentHeader[12:], "giztoy")                     // Vendor name
	binary.LittleEndian.PutUint32(commentHeader[18:], 0)  // User comment list length

	page = w.createPage(commentHeader, pageHeaderTypeContinuation, 0, stream.pageIndex, stream.serialNo)
	if _, err := w.w.Write(page); err != nil {
		return err
	}
	stream.pageIndex++

	return nil
}

// Write writes an Opus frame to the default stream.
func (w *OpusWriter) Write(frame opus.Frame) error {
	return w.StreamWrite(w.defaultStream, frame)
}

// StreamWrite writes an Opus frame to the specified stream.
func (w *OpusWriter) StreamWrite(serialNo int32, frame opus.Frame) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.closed {
		return ErrWriterClosed
	}

	stream, ok := w.streams[serialNo]
	if !ok {
		return ErrInvalidSerialNo
	}
	if stream.ended {
		return ErrStreamEnded
	}

	// Calculate granule increment from frame duration (always at 48kHz for OGG Opus)
	samples := int64(frame.Duration() * 48000 / time.Second)
	stream.granule += samples

	page := w.createPage(frame, pageHeaderTypeContinuation, uint64(stream.granule), stream.pageIndex, stream.serialNo)
	if _, err := w.w.Write(page); err != nil {
		return err
	}
	stream.pageIndex++

	return nil
}

// Granule returns the current granule position of the default stream.
func (w *OpusWriter) Granule() int64 {
	return w.StreamGranule(w.defaultStream)
}

// StreamGranule returns the current granule position of the specified stream.
func (w *OpusWriter) StreamGranule(serialNo int32) int64 {
	w.mu.Lock()
	defer w.mu.Unlock()

	stream, ok := w.streams[serialNo]
	if !ok {
		return 0
	}
	return stream.granule
}

// SetGranule sets the granule position of the default stream.
func (w *OpusWriter) SetGranule(g int64) {
	w.StreamSetGranule(w.defaultStream, g)
}

// StreamSetGranule sets the granule position of the specified stream.
func (w *OpusWriter) StreamSetGranule(serialNo int32, g int64) {
	w.mu.Lock()
	defer w.mu.Unlock()

	stream, ok := w.streams[serialNo]
	if !ok {
		return
	}
	stream.granule = g
}

// StreamEnd ends the specified stream by writing an EOS page.
func (w *OpusWriter) StreamEnd(serialNo int32) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.closed {
		return ErrWriterClosed
	}

	stream, ok := w.streams[serialNo]
	if !ok {
		return ErrInvalidSerialNo
	}
	if stream.ended {
		return nil // Already ended
	}

	// Write an empty EOS page
	page := w.createPage(nil, pageHeaderTypeEOS, uint64(stream.granule), stream.pageIndex, stream.serialNo)
	if _, err := w.w.Write(page); err != nil {
		return err
	}
	stream.ended = true

	return nil
}

// Close ends all streams and closes the writer.
func (w *OpusWriter) Close() error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.closed {
		return nil
	}

	// End all streams that haven't been ended yet
	for serialNo, stream := range w.streams {
		if !stream.ended {
			page := w.createPage(nil, pageHeaderTypeEOS, uint64(stream.granule), stream.pageIndex, serialNo)
			w.w.Write(page) // Ignore error on close
			stream.ended = true
		}
	}

	w.closed = true

	// Close underlying writer if it implements io.Closer
	if closer, ok := w.w.(io.Closer); ok {
		return closer.Close()
	}

	return nil
}

// createPage creates an OGG page with the given payload.
func (w *OpusWriter) createPage(payload []byte, headerType uint8, granulePos uint64, pageIndex uint32, serialNo int32) []byte {
	payloadLen := len(payload)
	nSegments := 1
	if payloadLen > 0 {
		nSegments = (payloadLen / 255) + 1
	}

	page := make([]byte, pageHeaderSize+nSegments+payloadLen)

	copy(page[0:], pageHeaderSignature)                     // "OggS"
	page[4] = 0                                             // Version
	page[5] = headerType                                    // Header type
	binary.LittleEndian.PutUint64(page[6:], granulePos)     // Granule position
	binary.LittleEndian.PutUint32(page[14:], uint32(serialNo)) // Serial number
	binary.LittleEndian.PutUint32(page[18:], pageIndex)     // Page sequence number
	// page[22:26] is checksum, filled later
	page[26] = uint8(nSegments)                             // Number of segments

	// Fill segment table
	if payloadLen > 0 {
		for i := 0; i < nSegments-1; i++ {
			page[pageHeaderSize+i] = 255
		}
		page[pageHeaderSize+nSegments-1] = uint8(payloadLen % 255)
	} else {
		page[pageHeaderSize] = 0
	}

	// Copy payload
	copy(page[pageHeaderSize+nSegments:], payload)

	// Calculate and set checksum
	var checksum uint32
	for i := range page {
		checksum = (checksum << 8) ^ w.checksumTable[byte(checksum>>24)^page[i]]
	}
	binary.LittleEndian.PutUint32(page[22:], checksum)

	return page
}

// generateChecksumTable generates the CRC32 lookup table for OGG.
func generateChecksumTable() *[256]uint32 {
	var table [256]uint32
	const poly = 0x04c11db7

	for i := range table {
		r := uint32(i) << 24
		for j := 0; j < 8; j++ {
			if (r & 0x80000000) != 0 {
				r = (r << 1) ^ poly
			} else {
				r <<= 1
			}
			table[i] = r & 0xffffffff
		}
	}
	return &table
}

File: rust/audio/src/codec/ogg/opus_writer.rs
//! Opus-in-OGG writer for multi-stream Opus audio.
//!
//! Supports writing multiple Opus streams with different serial numbers
//! into a single OGG container, with proper OpusHead/OpusTags headers
//! and granule position tracking.

use std::collections::HashMap;
use std::io::{self, Write};

/// Normal data page (no special flags). OGG spec: 0x00 = fresh packet.
/// Note: OGG continuation flag is 0x01, which is different.
const PAGE_HEADER_TYPE_FRESH: u8 = 0x00;
const PAGE_HEADER_TYPE_BOS: u8 = 0x02;
const PAGE_HEADER_TYPE_EOS: u8 = 0x04;
const DEFAULT_PRE_SKIP: u16 = 3840; // 80ms at 48kHz (RFC 7845 §5.1)
const PAGE_HEADER_SIGNATURE: &[u8] = b"OggS";
const PAGE_HEADER_SIZE: usize = 27;

/// A single Opus stream within the OGG container.
struct OpusStream {
    serial_no: i32,
    sample_rate: u32,
    channels: u16,
    granule: i64,
    page_index: u32,
    ended: bool,
    tags_written: bool,
}

/// Writes Opus frames into an OGG container.
///
/// Supports multiple streams with different serial numbers.
pub struct OpusWriter<W: Write> {
    writer: W,
    streams: HashMap<i32, OpusStream>,
    default_stream: i32,
    checksum_table: [u32; 256],
    closed: bool,
}

impl<W: Write> OpusWriter<W> {
    /// Creates a new OGG Opus writer with a default stream.
    pub fn new(writer: W, sample_rate: u32, channels: u16) -> io::Result<Self> {
        let mut ow = Self {
            writer,
            streams: HashMap::new(),
            default_stream: 0,
            checksum_table: generate_checksum_table(),
            closed: false,
        };

        let serial_no = ow.stream_begin(sample_rate, channels)?;
        ow.default_stream = serial_no;
        Ok(ow)
    }

    /// Creates a new stream and writes its BOS (OpusHead) page.
    ///
    /// Per RFC 3533, all BOS pages must appear before any non-BOS pages.
    /// The OpusTags page is deferred until the first `stream_write` call.
    /// Returns the serial number assigned to this stream.
    pub fn stream_begin(&mut self, sample_rate: u32, channels: u16) -> io::Result<i32> {
        let serial_no = self.generate_serial_no();
        let mut stream = OpusStream {
            serial_no,
            sample_rate,
            channels,
            granule: 0,
            page_index: 0,
            ended: false,
            tags_written: false,
        };

        self.write_bos_page(&mut stream)?;
        self.streams.insert(serial_no, stream);
        Ok(serial_no)
    }

    fn generate_serial_no(&self) -> i32 {
        // Simple incrementing serial for deterministic behavior
        (self.streams.len() as i32) + 1
    }

    fn write_bos_page(&mut self, stream: &mut OpusStream) -> io::Result<()> {
        let mut id_header = vec![0u8; 19];
        id_header[..8].copy_from_slice(b"OpusHead");
        id_header[8] = 1; // Version
        id_header[9] = stream.channels as u8;
        id_header[10..12].copy_from_slice(&DEFAULT_PRE_SKIP.to_le_bytes());
        id_header[12..16].copy_from_slice(&stream.sample_rate.to_le_bytes());
        id_header[16..18].copy_from_slice(&0u16.to_le_bytes()); // Output gain
        id_header[18] = 0; // Channel mapping

        let page = self.create_page(&id_header, PAGE_HEADER_TYPE_BOS, 0, stream.page_index, stream.serial_no);
        self.writer.write_all(&page)?;
        stream.page_index += 1;
        Ok(())
    }

    fn write_tags_page(&mut self, stream: &mut OpusStream) -> io::Result<()> {
        let mut comment_header = vec![0u8; 22];
        comment_header[..8].copy_from_slice(b"OpusTags");
        comment_header[8..12].copy_from_slice(&6u32.to_le_bytes()); // Vendor length
        comment_header[12..18].copy_from_slice(b"giztoy"); // Vendor
        comment_header[18..22].copy_from_slice(&0u32.to_le_bytes()); // No comments

        let page = self.create_page(&comment_header, PAGE_HEADER_TYPE_FRESH, 0, stream.page_index, stream.serial_no);
        self.writer.write_all(&page)?;
        stream.page_index += 1;
        Ok(())
    }

    /// Writes an Opus frame to the default stream.
    ///
    /// `duration_48k` is the frame duration in samples at 48kHz
    /// (e.g., 960 for 20ms, 480 for 10ms).
    pub fn write_frame(&mut self, frame: &[u8], duration_48k: i64) -> io::Result<()> {
        self.stream_write(self.default_stream, frame, duration_48k)
    }

    /// Writes an Opus frame to the specified stream.
    pub fn stream_write(&mut self, serial_no: i32, frame: &[u8], duration_48k: i64) -> io::Result<()> {
        if self.closed {
            return Err(io::Error::new(io::ErrorKind::BrokenPipe, "writer closed"));
        }

        let stream = self.streams.get_mut(&serial_no)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "invalid serial number"))?;

        if stream.ended {
            return Err(io::Error::new(io::ErrorKind::BrokenPipe, "stream ended"));
        }

        // Write OpusTags page on first data write (deferred from stream_begin
        // to ensure all BOS pages precede any non-BOS pages per RFC 3533).
        if !stream.tags_written {
            stream.tags_written = true;
            self.write_tags_page_for(serial_no)?;
            // Re-borrow after mutable self call
            let stream = self.streams.get_mut(&serial_no).unwrap();
            stream.granule += duration_48k;
            let granule = stream.granule as u64;
            let page_index = stream.page_index;
            stream.page_index += 1;

            let page = self.create_page(frame, PAGE_HEADER_TYPE_FRESH, granule, page_index, serial_no);
            return self.writer.write_all(&page);
        }

        stream.granule += duration_48k;
        let granule = stream.granule as u64;
        let page_index = stream.page_index;
        stream.page_index += 1;

        let page = self.create_page(frame, PAGE_HEADER_TYPE_FRESH, granule, page_index, serial_no);
        self.writer.write_all(&page)
    }

    /// Write tags page for a specific stream.
    fn write_tags_page_for(&mut self, serial_no: i32) -> io::Result<()> {
        let stream = self.streams.get(&serial_no).unwrap();
        let page_index = stream.page_index;

        let mut comment_header = vec![0u8; 22];
        comment_header[..8].copy_from_slice(b"OpusTags");
        comment_header[8..12].copy_from_slice(&6u32.to_le_bytes());
        comment_header[12..18].copy_from_slice(b"giztoy");
        comment_header[18..22].copy_from_slice(&0u32.to_le_bytes());

        let page = self.create_page(&comment_header, PAGE_HEADER_TYPE_FRESH, 0, page_index, serial_no);
        self.writer.write_all(&page)?;

        let stream = self.streams.get_mut(&serial_no).unwrap();
        stream.page_index += 1;
        Ok(())
    }

    /// Returns the current granule position of the default stream.
    pub fn granule(&self) -> i64 {
        self.streams.get(&self.default_stream).map(|s| s.granule).unwrap_or(0)
    }

    /// Ends the specified stream by writing an EOS page.
    pub fn stream_end(&mut self, serial_no: i32) -> io::Result<()> {
        if self.closed {
            return Err(io::Error::new(io::ErrorKind::BrokenPipe, "writer closed"));
        }

        let stream = self.streams.get_mut(&serial_no)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "invalid serial number"))?;

        if stream.ended {
            return Ok(());
        }

        let granule = stream.granule as u64;
        let page_index = stream.page_index;
        stream.ended = true;

        let page = self.create_page(&[], PAGE_HEADER_TYPE_EOS, granule, page_index, serial_no);
        self.writer.write_all(&page)
    }

    /// Closes the writer, ending all active streams.
    pub fn close(mut self) -> io::Result<()> {
        if self.closed {
            return Ok(());
        }

        let serial_nos: Vec<i32> = self.streams.keys().copied().collect();
        for serial_no in serial_nos {
            let _ = self.stream_end(serial_no);
        }
        self.closed = true;
        Ok(())
    }

    fn create_page(&self, payload: &[u8], header_type: u8, granule_pos: u64, page_index: u32, serial_no: i32) -> Vec<u8> {
        let payload_len = payload.len();
        let n_segments = if payload_len > 0 { (payload_len / 255) + 1 } else { 1 };

        let mut page = vec![0u8; PAGE_HEADER_SIZE + n_segments + payload_len];

        page[..4].copy_from_slice(PAGE_HEADER_SIGNATURE);
        page[4] = 0; // Version
        page[5] = header_type;
        page[6..14].copy_from_slice(&granule_pos.to_le_bytes());
        page[14..18].copy_from_slice(&(serial_no as u32).to_le_bytes());
        page[18..22].copy_from_slice(&page_index.to_le_bytes());
        // page[22..26] = checksum (filled later)
        page[26] = n_segments as u8;

        // Segment table
        if payload_len > 0 {
            for i in 0..n_segments - 1 {
                page[PAGE_HEADER_SIZE + i] = 255;
            }
            page[PAGE_HEADER_SIZE + n_segments - 1] = (payload_len % 255) as u8;
        } else {
            page[PAGE_HEADER_SIZE] = 0;
        }

        // Payload
        page[PAGE_HEADER_SIZE + n_segments..].copy_from_slice(payload);

        // Checksum
        let mut checksum: u32 = 0;
        for &b in &page {
            checksum = (checksum << 8) ^ self.checksum_table[((checksum >> 24) as u8 ^ b) as usize];
        }
        page[22..26].copy_from_slice(&checksum.to_le_bytes());

        page
    }
}

/// An Opus packet read from an OGG container.
#[derive(Debug, Clone)]
pub struct OpusPacket {
    /// Raw Opus frame data.
    pub data: Vec<u8>,
    /// Granule position.
    pub granule: i64,
    /// Stream serial number.
    pub serial_no: i32,
}

fn generate_checksum_table() -> [u32; 256] {
    let mut table = [0u32; 256];
    const POLY: u32 = 0x04c11db7;

    for i in 0..256 {
        let mut r = (i as u32) << 24;
        for _ in 0..8 {
            if (r & 0x80000000) != 0 {
                r = (r << 1) ^ POLY;
            } else {
                r <<= 1;
            }
        }
        table[i] = r;
    }
    table
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_opus_writer_creates_valid_ogg() {
        let mut buf = Vec::new();
        let mut writer = OpusWriter::new(&mut buf, 48000, 1).unwrap();

        // Write a fake opus frame (20ms at 48kHz = 960 samples)
        writer.write_frame(&[0xFC, 0x01, 0x02], 960).unwrap();
        writer.write_frame(&[0xFC, 0x03, 0x04], 960).unwrap();

        assert_eq!(writer.granule(), 1920);

        writer.close().unwrap();

        // Verify OGG header signature
        assert!(buf.len() > 4);
        assert_eq!(&buf[..4], b"OggS");
    }

    #[test]
    fn test_multi_stream() {
        let mut buf = Vec::new();
        let mut writer = OpusWriter::new(&mut buf, 48000, 1).unwrap();

        let stream2 = writer.stream_begin(48000, 2).unwrap();

        writer.write_frame(&[0xFC, 0x01], 960).unwrap();
        writer.stream_write(stream2, &[0xFC, 0x02], 960).unwrap();

        writer.stream_end(stream2).unwrap();
        writer.close().unwrap();

        assert!(buf.len() > 100);
    }

    #[test]
    fn test_checksum_table() {
        let table = generate_checksum_table();
        assert_eq!(table[0], 0);
        assert_ne!(table[1], 0);
    }
}
